modelskill/matching.py

```diff
@@ -209,14 +214,230 @@ def _single_obs_compare(
+    return df.loc[valid_time]
+
+
+def _get_valid_query_time(
+    mod_index: pd.DatetimeIndex, obs_index: pd.DatetimeIndex, max_gap: pd.Timedelta
+):
+    """Used only by _remove_model_gaps"""
+    # init dataframe of available timesteps and their index
+    df = pd.DataFrame(index=mod_index)
+    df["idx"] = range(len(df))
+
+    # for query times get available left and right index of source times
+    df = _interp_time(df, obs_index).dropna()
+    df["idxa"] = np.floor(df.idx).astype(int)
+    df["idxb"] = np.ceil(df.idx).astype(int)
+
+    # time of left and right source times and time delta
+    df["ta"] = mod_index[df.idxa]
+    df["tb"] = mod_index[df.idxb]
+    df["dt"] = df.tb - df.ta
+
+    # valid query times where time delta is less than max_gap
+    valid_idx = df.dt <= max_gap
+    return valid_idx
+
+
+def _mask_model_outside_observation_track(name, df_mod, df_obs) -> None:
+    if len(df_mod) == 0:
+        return
+    if len(df_mod) != len(df_obs):
+        raise ValueError("model and observation data must have same length")
+
+    mod_xy = df_mod[["x", "y"]]
+    obs_xy = df_obs[["x", "y"]]
+    d_xy = np.sqrt(np.sum((obs_xy - mod_xy) ** 2, axis=1))
+
+    # Find minimal accepted distance from data
+    # (could be long/lat or x/y, small or large domain)
+    tol_xy = _minimal_accepted_distance(obs_xy)
+    mask = d_xy > tol_xy
+    df_mod.loc[mask, name] = np.nan
+    if all(mask):
+        warnings.warn("no (spatial) overlap between model and observation points")
+
+
+def _get_model_start_end(raw_mod_data) -> Tuple[pd.Timestamp, pd.Timestamp]:
+    """Find first start and last end time of all model data"""
+    _mod_start = [pd.Timestamp.max]
+    _mod_end = [pd.Timestamp.min]
+    for m in raw_mod_data.values():
+        if len(m) == 0:
+            continue
+        _mod_start.append(m.index[0])  # TODO: xr.Dataset
+        _mod_end.append(m.index[-1])  # TODO: xr.Dataset
+    return min(_mod_start), max(_mod_end)
+
+
+def match_time(
+    observation: Observation,
+    raw_mod_data: Dict[str, pd.DataFrame],
+    max_model_gap: Optional[TimeDeltaTypes] = None,
+) -> xr.Dataset:
+    """Match observation with one or more model results in time domain
+    and return as xr.Dataset in the format used by modelskill.Comparer
+
+    Will interpolate model results to observation time.
+
+    Note: assumes that observation and model data are already matched in space.
+
+    Parameters
+    ----------
+    observation : Observation
+        Observation to be matched
+    raw_mod_data : Dict[str, pd.DataFrame]
+        Dictionary of model results ready for interpolation
+    max_model_gap : Optional[TimeDeltaTypes], optional
+        In case of non-equidistant model results (e.g. event data),
+        max_model_gap can be given e.g. as seconds, by default None
+
+    Returns
+    -------
+    xr.Dataset
+        Matched data in the format used by modelskill.Comparer
+    """
+    _obs_name = "Observation"
```

Comment:
What is the intention with the underscore prefix for these local variables? Private local variables 🤔